# Парсинг и предобработка корпуса

In [1]:
import pandas as pd
from string import punctuation as punct
from swda import Transcript, CorpusReader
from tqdm.auto import tqdm
import joblib
import os

pd.set_option('display.max_colwidth', -1)

In [2]:
metadata = pd.read_csv("swda/swda-metadata.csv")

In [3]:
metadata.head()

conversation_no  talk_day  length    topic_description  \
0  1001             910207    7       WEATHER CLIMATE       
1  1002             910207    10      CARE OF THE ELDERLY   
2  1003             910207    7       VACATION SPOTS        
3  1005             910207    2       CARE OF THE ELDERLY   
4  1006             910207    5       WEATHER CLIMATE       

                                                                                                                                                                                                                   prompt  \
0  DISCUSS THE WEATHER.   WHAT HAS IT BEEN LIKE IN YOUR AREA?   HAS IT BEEN TYPICAL FOR THIS TIME OF YEAR?  COMPARE IT WITH THE OTHER CALLER'S WEATHER.                                                                     
1  PLEASE DISCUSS CARE OF THE ELDERLY.  FIND OUT HOW THE OTHER CALLER FEELS ABOUT SENDING AN ELDERLY FAMILY MEMBER TO A NURSING HOME.  WHAT SHOULD ONE KNOW ABOUT THE NURSING HOME ENVIRONMENT WHEN MAKING THIS DECISION?   
2  PLEASE DISCUSS TYPES OF VACATIONS AND TRIPS YOU ENJOY.  FIND OUT WHETHER THE OTHER CALLER CAN INTEREST YOU IN A VACATION SPOT YOU HAVEN'T VISITED.                                                                       
3  PLEASE DISCUSS CARE OF THE ELDERLY.  FIND OUT HOW THE OTHER CALLER FEELS ABOUT SENDING AN ELDERLY FAMILY MEMBER TO A NURSING HOME.  WHAT SHOULD ONE KNOW ABOUT THE NURSING HOME ENVIRONMENT WHEN MAKING THIS DECISION?   
4  DISCUSS THE WEATHER.   WHAT HAS IT BEEN LIKE IN YOUR AREA?   HAS IT BEEN TYPICAL FOR THIS TIME OF YEAR?  COMPARE IT WITH THE OTHER CALLER'S WEATHER.                                                                     

   from_caller from_caller_sex  from_caller_education  from_caller_birth_year  \
0  1165         MALE            2                      1937                     
1  1195         MALE            0                      1947                     
2  1196         MALE            0                      1940                     
3  1165         MALE            2                      1937                     
4  1190         MALE            3                      1941                     

  from_caller_dialect_area  to_caller to_caller_sex  to_caller_education  \
0  NORTHERN                 1065       MALE          3                     
1  NORTHERN                 1037       MALE          3                     
2  NYC                      1192       MALE          0                     
3  NORTHERN                 1203       MALE          0                     
4  NYC                      1198       MALE          0                     

   to_caller_birth_year to_caller_dialect_area  
0  1952                  WESTERN                
1  1947                  WESTERN                
2  1957                  NORTHERN               
3  1961                  NORTH MIDLAND          
4  1950                  NYC

In [4]:
metadata.conversation_no.nunique()

2866

In [5]:
metadata.prompt.nunique()

66

## Вроде бы это означает, что в данных у нас 2866 диалогов

In [7]:
metadata.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2866 entries, 0 to 2865
Data columns (total 15 columns):
conversation_no             2866 non-null int64
talk_day                    2866 non-null int64
length                      2866 non-null int64
topic_description           2866 non-null object
prompt                      2866 non-null object
from_caller                 2866 non-null int64
from_caller_sex             2866 non-null object
from_caller_education       2866 non-null int64
from_caller_birth_year      2866 non-null int64
from_caller_dialect_area    2861 non-null object
to_caller                   2866 non-null int64
to_caller_sex               2866 non-null object
to_caller_education         2866 non-null int64
to_caller_birth_year        2866 non-null int64
to_caller_dialect_area      2862 non-null object
dtypes: int64(9), object(6)
memory usage: 335.9+ KB


In [8]:
print(len(metadata.topic_description.unique()))
metadata.topic_description.unique()

66


array(['WEATHER CLIMATE', 'CARE OF THE ELDERLY', 'VACATION SPOTS',
       'GUN CONTROL', 'AIR POLLUTION', 'MUSIC', 'UNIVERSAL PBLIC SERV',
       'CRIME', 'CREDIT CARD USE', 'AIDS', 'BOOKS AND LITERATURE',
       'CLOTHING AND DRESS', 'CAMPING', 'MOVIES', 'RIGHT TO PRIVACY',
       'TAXES', 'HOUSES', "WOMEN'S ROLES", 'PETS', 'FAMILY FINANCE',
       'TRIAL BY JURY', 'FAMILY LIFE', 'PUERTO RICAN STTEHD',
       'LATIN AMERICA', 'TV PROGRAMS', 'EXERCISE AND FITNESS',
       'HOBBIES AND CRAFTS', 'CHILD CARE', 'RECIPES/FOOD/COOKING',
       'NEWS MEDIA', 'JOB BENEFITS', 'CAPITAL PUNISHMENT', 'BUYING A CAR',
       'ELECTIONS AND VOTING', 'BASEBALL', 'FISHING', 'DRUG TESTING',
       'SOVIET UNION', 'RECYCLING', 'FOOTBALL', 'COMPUTERS',
       'RESTAURANTS', 'POLITICS', 'CHOOSING A COLLEGE', 'BASKETBALL',
       'PUBLIC EDUCATION', 'GARDENING', 'PAINTING', 'SVGS & LOAN BAILOUT',
       'VIETNAM WAR', 'FEDERAL BUDGET', 'MIDDLE EAST', 'IMMIGRATION',
       'UNIVERSAL HEALTH INS', 'AUTO REPAI

In [9]:
corpus = CorpusReader('swda')

Поля реплик: 
* caller	str	A, B, @A, @B, @@A, @@B
* caller_no	int	The caller Id.
* caller_sex	str	MALE or FEMALE
* caller_education	str	0, 1, 2, 3, 9
* caller_birth_year	int	4-digit year
* caller_dialect_area	str	MIXED, NEW ENGLAND, NORTH MIDLAND, NORTHERN, NYC, SOUTH MIDLAND, SOUTHERN, UNK, WESTERN
* transcript_index	int	line number relative to the whole transcript
* utterance_index	int	Utterance number (can span multiple TranscriptIndex numbers)
* subutterance_Index	int	Utterances can be broken across line. This gives the internal position.
* act_tag	list	strings; see below
* text	str	the text of the utterance
* pos	str	the part-of-speech tagged portion of the utterance
* trees	nltk.tree.Tree	the parse of Text; see below for discussion

In [10]:
def clever_string_join(words: list):
    
    res = []
    for utt in words:
        string = ""
        for tok in utt:
            if "'" in tok or tok in punct:
                string += tok
            else: 
                string += " " + tok
        res.append(string.strip())
        
    return res

In [13]:
def create_df(corpus):
        
    talk_indx = []
    callers = []
    utt_indx = []
    trans_indx = []
    texts = []
    tags = []
    words = []
    topics = []
    prompts = []

    for dialogue in corpus.iter_transcripts():
        
        for utt in dialogue.utterances:
            texts.append(utt.text)
            callers.append(utt.caller)
            tags.append(utt.damsl_act_tag())
            utt_indx.append(utt.utterance_index)
            trans_indx.append(utt.transcript_index)
            words.append(utt.pos_words())
            talk_indx.append(dialogue.conversation_no)
            topics.append(dialogue.topic_description)
            prompts.append(dialogue.prompt)
            
    new_words = clever_string_join(words)
        
    d = {"dialogue": talk_indx,
        "caller":callers, 
        "utt_indx": utt_indx, 
        "trans_indx": trans_indx,
        "orig_text": texts, 
        "tag": tags,
        "clean_text": new_words,
        "topic": topics,
        "prompt": prompts}
    
    return d

In [46]:
d = create_df(corpus) # несколько минут

transcript 1155


In [147]:
df = pd.DataFrame(d)
df.head(40)

dialogue caller  utt_indx  trans_indx  \
0   3774      A      1         0            
1   3774      B      2         1            
2   3774      A      3         2            
3   3774      B      4         3            
4   3774      A      5         4            
5   3774      A      5         5            
6   3774      B      6         6            
7   3774      A      7         7            
8   3774      B      8         8            
9   3774      A      9         9            
10  3774      B      10        10           
11  3774      B      10        11           
12  3774      B      10        12           
13  3774      B      10        13           
14  3774      A      11        14           
15  3774      A      11        15           
16  3774      B      12        16           
17  3774      B      12        17           
18  3774      B      12        18           
19  3774      B      12        19           
20  3774      A      13        20           
21  3774      B      14        21           
22  3774      A      15        22           
23  3774      A      15        23           
24  3774      B      16        24           
25  3774      A      17        25           
26  3774      A      17        26           
27  3774      B      18        27           
28  3774      A      19        28           
29  3774      A      19        29           
30  3774      B      20        30           
31  3774      A      21        31           
32  3774      B      22        32           
33  3774      A      23        33           
34  3774      B      24        34           
35  3774      A      25        35           
36  3774      B      26        37           
37  3774      A      27        38           
38  3774      B      28        39           
39  3774      A      29        40           

                                                                                                                      orig_text  \
0   {C So } I've been concerned about crime lately. /                                                                             
1   Uh-huh. /                                                                                                                     
2   {F Uh, } it's really scary to listen to the news every night  and --                                                          
3   Uh-huh. /                                                                                                                     
4   -- to hear about all the problems.  /                                                                                         
5   I wondered if you were taking any special precautions in your neighborhood? /                                                 
6   {D Well, } [ I, + I  ] think we have a neighborhood watch <laughter>. /                                                       
7   Uh-huh. /                                                                                                                     
8   I think. /                                                                                                                    
9   <Laughter>.                                                                                                                   
10  I'm not real, - /                                                                                                             
11  we don't get real involved.  /                                                                                                
12  We're never home,  /                                                                                                          
13  {C so } -- -/                                                                                                                 
14  Uh-huh  /                                                                                                                     
15  <laughter>.                                               

# Попробуем посмотреть распределение каждого тега

In [148]:
len(df)

221616

In [149]:
df["extra"] = 1

In [150]:
tag_stats = df.groupby(by="tag", as_index=False).extra.sum()
tag_stats

tag  extra
0   %                15550
1   +                17824
2   ^2               723  
3   ^g               92   
4   ^h               556  
5   ^q               983  
6   aa               11133
7   aap_am           105  
8   ad               746  
9   ar               345  
10  arp_nd           207  
11  b                38298
12  b^m              688  
13  ba               4764 
14  bd               103  
15  bf               952  
16  bh               1053 
17  bk               1306 
18  br               298  
19  fa               79   
20  fc               2582 
21  fo_o_fw_"_by_bc  883  
22  fp               225  
23  ft               78   
24  h                1226 
25  na               847  
26  ng               302  
27  nn               1377 
28  no               285  
29  ny               3034 
30  oo_co_cc         110  
31  qh               575  
32  qo               656  
33  qrr              209  
34  qw               1979 
35  qw^d             80   
36  qy               4726 
37  qy^d             1218 
38  sd               75143
39  sv               26426
40  t1               103  
41  t3               117  
42  x                3630

In [151]:
tag_stats["extra"] = tag_stats.extra.apply(lambda x: round(x/221616,2))
tag_stats.sort_values(by="extra", ascending=False)

tag  extra
38  sd               0.34 
11  b                0.17 
39  sv               0.12 
1   +                0.08 
0   %                0.07 
6   aa               0.05 
36  qy               0.02 
13  ba               0.02 
42  x                0.02 
27  nn               0.01 
20  fc               0.01 
29  ny               0.01 
24  h                0.01 
37  qy^d             0.01 
17  bk               0.01 
34  qw               0.01 
30  oo_co_cc         0.00 
31  qh               0.00 
32  qo               0.00 
33  qrr              0.00 
9   ar               0.00 
28  no               0.00 
4   ^h               0.00 
3   ^g               0.00 
2   ^2               0.00 
40  t1               0.00 
41  t3               0.00 
35  qw^d             0.00 
26  ng               0.00 
10  arp_nd           0.00 
25  na               0.00 
23  ft               0.00 
22  fp               0.00 
5   ^q               0.00 
19  fa               0.00 
18  br               0.00 
16  bh               0.00 
15  bf               0.00 
14  bd               0.00 
7   aap_am           0.00 
12  b^m              0.00 
8   ad               0.00 
21  fo_o_fw_"_by_bc  0.00

In [152]:
# смотрим, какие предложения есть у каждого тега
df[df.tag == "%"].clean_text

10        I'm not real,                  
13        so --                          
50        and,                           
58        so --                          
61        so,                            
69        that's --                      
71        -- what it is.                 
79        and they don't, uh,            
86        and, course, you know,         
87        I don't know,                  
116       Really is.                     
126       Um.                            
163       Um.                            
182       so,                            
194       and yet there's,               
199       so,                            
210       -- and, uh, so,                
222       Uh, but we,                    
229       So,                            
236       Uh, but, uh, anyway, but, okay.
256       so,                            
261       So it's,                       
269       Uh, but, uh,                   
277       Uh-,                    

## Теперь избавимся от фраз с тегом "b", которые прерывают текущего говорящего (т.е. когда предыдущая фраза оканчивается на "--")

In [153]:
df.head()

dialogue caller  utt_indx  trans_indx  \
0  3774      A      1         0            
1  3774      B      2         1            
2  3774      A      3         2            
3  3774      B      4         3            
4  3774      A      5         4            

                                                              orig_text tag  \
0  {C So } I've been concerned about crime lately. /                     sd   
1  Uh-huh. /                                                             b    
2  {F Uh, } it's really scary to listen to the news every night  and --  sv   
3  Uh-huh. /                                                             b    
4  -- to hear about all the problems.  /                                 +    

                                                       clean_text  topic  \
0  So I've been concerned about crime lately.                      CRIME   
1  Uh-huh.                                                         CRIME   
2  Uh, it's really scary to listen to the news every night and --  CRIME   
3  Uh-huh.                                                         CRIME   
4  -- to hear about all the problems.                              CRIME   

                                                                                                                                            prompt  \
0  DISCUSS CRIME IN AMERICAN CITIES TODAY.  WHAT ARE YOUR CONCERNS AND THE CONCERNS OF THE OTHER CALLER?  WHAT STEPS CAN BE TAKEN TO REDUCE CRIME?   
1  DISCUSS CRIME IN AMERICAN CITIES TODAY.  WHAT ARE YOUR CONCERNS AND THE CONCERNS OF THE OTHER CALLER?  WHAT STEPS CAN BE TAKEN TO REDUCE CRIME?   
2  DISCUSS CRIME IN AMERICAN CITIES TODAY.  WHAT ARE YOUR CONCERNS AND THE CONCERNS OF THE OTHER CALLER?  WHAT STEPS CAN BE TAKEN TO REDUCE CRIME?   
3  DISCUSS CRIME IN AMERICAN CITIES TODAY.  WHAT ARE YOUR CONCERNS AND THE CONCERNS OF THE OTHER CALLER?  WHAT STEPS CAN BE TAKEN TO REDUCE CRIME?   
4  DISCUSS CRIME IN AMERICAN CITIES TODAY.  WHAT ARE YOUR CONCERNS AND THE CONCERNS OF THE OTHER CALLER?  WHAT STEPS CAN BE TAKEN TO REDUCE CRIME?   

   extra  
0  1      
1  1      
2  1      
3  1      
4  1

In [154]:
to_delete = []
counter = 0

for i in df.index:
    if df.loc[i, "tag"] == "b" and df.loc[i-1, "orig_text"].endswith("--"):
        counter += 1
        to_delete.append(i)

In [155]:
counter

5284

In [156]:
to_delete[:20]

[3,
 24,
 27,
 30,
 38,
 40,
 43,
 81,
 90,
 92,
 98,
 110,
 130,
 142,
 158,
 177,
 190,
 197,
 207,
 215]

In [157]:
df.iloc[22:26, :]

dialogue caller  utt_indx  trans_indx  \
22  3774      A      15        22           
23  3774      A      15        23           
24  3774      B      16        24           
25  3774      A      17        25           

                                                                                                       orig_text  \
22  {D Well } <breathing>, [ we moved in, + when we moved in,  ] [ there, + there ] wasn't any outside lights  /   
23  {C and --                                                                                                      
24  Uh-huh. /                                                                                                      
25  -- so } we've been trying to install some, {F uh, } outside lights  /                                          

   tag  \
22  sd   
23  sd   
24  b    
25  +    

                                                                     clean_text  \
22  Well, we moved in, when we moved in, there, there wasn't any outside lights   
23  and --                                                                        
24  Uh-huh.                                                                       
25  -- so we've been trying to install some, uh, outside lights                   

    topic  \
22  CRIME   
23  CRIME   
24  CRIME   
25  CRIME   

                                                                                                                                             prompt  \
22  DISCUSS CRIME IN AMERICAN CITIES TODAY.  WHAT ARE YOUR CONCERNS AND THE CONCERNS OF THE OTHER CALLER?  WHAT STEPS CAN BE TAKEN TO REDUCE CRIME?   
23  DISCUSS CRIME IN AMERICAN CITIES TODAY.  WHAT ARE YOUR CONCERNS AND THE CONCERNS OF THE OTHER CALLER?  WHAT STEPS CAN BE TAKEN TO REDUCE CRIME?   
24  DISCUSS CRIME IN AMERICAN CITIES TODAY.  WHAT ARE YOUR CONCERNS AND THE CONCERNS OF THE OTHER CALLER?  WHAT STEPS CAN BE TAKEN TO REDUCE CRIME?   
25  DISCUSS CRIME IN AMERICAN CITIES TODAY.  WHAT ARE YOUR CONCERNS AND THE CONCERNS OF THE OTHER CALLER?  WHAT STEPS CAN BE TAKEN TO REDUCE CRIME?   

    extra  
22  1      
23  1      
24  1      
25  1

In [158]:
df.iloc[195:200, :]

dialogue caller  utt_indx  trans_indx  \
195  3184      A      11        26           
196  3184      A      11        27           
197  3184      B      12        28           
198  3184      A      13        29           
199  3184      A      13        30           

                                                                                                                                               orig_text  \
195  where we can view the river from that distance, there's still homes within our area that are, {D you know, } pretty typical of what our home is,  /   
196  {C so } --                                                                                                                                            
197  Uh-huh /                                                                                                                                              
198  -- it's, {F uh, } pretty much that, {F uh, }  type of, {F uh, } home  /                                                                               
199  {C so, } -/                                                                                                                                           

    tag  \
195  sd   
196  sd   
197  b    
198  +    
199  %    

                                                                                                                                      clean_text  \
195  where we can view the river from that distance, there's still homes within our area that are, you know, pretty typical of what our home is,   
196  so --                                                                                                                                         
197  Uh-huh                                                                                                                                        
198  -- it's, uh, pretty much that, uh, type of, uh, home                                                                                          
199  so,                                                                                                                                           

      topic  \
195  HOUSES   
196  HOUSES   
197  HOUSES   
198  HOUSES   
199  HOUSES   

                                                                                                             prompt  \
195  FIND OUT ABOUT THE OTHER CALLER'S HOME.  IS IT A TYPICAL HOME FOR THE AREA?  HOW DOES IT COMPARE TO YOUR HOME?   
196  FIND OUT ABOUT THE OTHER CALLER'S HOME.  IS IT A TYPICAL HOME FOR THE AREA?  HOW DOES IT COMPARE TO YOUR HOME?   
197  FIND OUT ABOUT THE OTHER CALLER'S HOME.  IS IT A TYPICAL HOME FOR THE AREA?  HOW DOES IT COMPARE TO YOUR HOME?   
198  FIND OUT ABOUT THE OTHER CALLER'S HOME.  IS IT A TYPICAL HOME FOR THE AREA?  HOW DOES IT COMPARE TO YOUR HOME?   
199  FIND OUT ABOUT THE OTHER CALLER'S HOME.  IS IT A TYPICAL HOME FOR THE AREA?  HOW DOES IT COMPARE TO YOUR HOME?   

     extra  
195  1      
196  1      
197  1      
198  1      
199  1

In [159]:
# вроде бы все в порядке - можно удалять

In [160]:
new_df = df.drop(to_delete)
new_df = new_df.reset_index(drop=True)

In [162]:
new_df.head()

dialogue caller  utt_indx  trans_indx  \
0  3774      A      1         0            
1  3774      B      2         1            
2  3774      A      3         2            
3  3774      A      5         4            
4  3774      A      5         5            

                                                                       orig_text  \
0  {C So } I've been concerned about crime lately. /                               
1  Uh-huh. /                                                                       
2  {F Uh, } it's really scary to listen to the news every night  and --            
3  -- to hear about all the problems.  /                                           
4  I wondered if you were taking any special precautions in your neighborhood? /   

    tag  \
0  sd     
1  b      
2  sv     
3  +      
4  qy^d   

                                                                    clean_text  \
0  So I've been concerned about crime lately.                                    
1  Uh-huh.                                                                       
2  Uh, it's really scary to listen to the news every night and --                
3  -- to hear about all the problems.                                            
4  I wondered if you were taking any special precautions in your neighborhood?   

   topic  \
0  CRIME   
1  CRIME   
2  CRIME   
3  CRIME   
4  CRIME   

                                                                                                                                            prompt  \
0  DISCUSS CRIME IN AMERICAN CITIES TODAY.  WHAT ARE YOUR CONCERNS AND THE CONCERNS OF THE OTHER CALLER?  WHAT STEPS CAN BE TAKEN TO REDUCE CRIME?   
1  DISCUSS CRIME IN AMERICAN CITIES TODAY.  WHAT ARE YOUR CONCERNS AND THE CONCERNS OF THE OTHER CALLER?  WHAT STEPS CAN BE TAKEN TO REDUCE CRIME?   
2  DISCUSS CRIME IN AMERICAN CITIES TODAY.  WHAT ARE YOUR CONCERNS AND THE CONCERNS OF THE OTHER CALLER?  WHAT STEPS CAN BE TAKEN TO REDUCE CRIME?   
3  DISCUSS CRIME IN AMERICAN CITIES TODAY.  WHAT ARE YOUR CONCERNS AND THE CONCERNS OF THE OTHER CALLER?  WHAT STEPS CAN BE TAKEN TO REDUCE CRIME?   
4  DISCUSS CRIME IN AMERICAN CITIES TODAY.  WHAT ARE YOUR CONCERNS AND THE CONCERNS OF THE OTHER CALLER?  WHAT STEPS CAN BE TAKEN TO REDUCE CRIME?   

   extra  
0  1      
1  1      
2  1      
3  1      
4  1

## Теперь объединим высказывания с тегом "+" с предыдущими высказываниями, если они относятся к тому же спикеру и кончаются на "--"

In [163]:
to_merge = []
counter = 0

for i in new_df.index:
    if (new_df.loc[i, "tag"] == "+" 
        and new_df.loc[i-1, "orig_text"].endswith("--")
       and new_df.loc[i-1, "caller"] == new_df.loc[i, "caller"]):
        counter += 1
        to_merge.append(i)

In [164]:
counter

5050

In [165]:
to_merge[-5:]

[215595, 215620, 216220, 216258, 216277]

In [166]:
new_df.iloc[215593:215597]

dialogue caller  utt_indx  trans_indx  \
215593  3018      A      15        26           
215594  3018      A      15        27           
215595  3018      A      17        29           
215596  3018      B      18        30           

                                                                                                                                        orig_text  \
215593  <inhaling>.  {C And } [ I, +  I ]  have watched THIRTYSOMETHING some in the last couple seasons.  /                                         
215594  I was sort of sorry to see it go --                                                                                                         
215595  -- too. /                                                                                                                                   
215596  [ It, + it ] was kind of interesting, {F uh, }  some people complained about, {F uh, } {F uh, } the kind of whining or whatever on it.  /   

       tag  \
215593  sd   
215594  sd   
215595  +    
215596  sd   

                                                                                                              clean_text  \
215593  . And I, I have watched THIRTYSOMETHING some in the last couple seasons.                                           
215594  I was sort of sorry to see it go --                                                                                
215595  -- too.                                                                                                            
215596  It, it was kind of interesting, uh, some people complained about, uh, uh, the kind of whining or whatever on it.   

              topic  \
215593  TV PROGRAMS   
215594  TV PROGRAMS   
215595  TV PROGRAMS   
215596  TV PROGRAMS   

                                                                                                           prompt  \
215593  FIND OUT WHAT THE OTHER CALLER'S FAVORITE TV SHOWS ARE AND WHY.  ARE YOUR INTERESTS SIMILAR OR DIFFERENT?   
215594  FIND OUT WHAT THE OTHER CALLER'S FAVORITE TV SHOWS ARE AND WHY.  ARE YOUR INTERESTS SIMILAR OR DIFFERENT?   
215595  FIND OUT WHAT THE OTHER CALLER'S FAVORITE TV SHOWS ARE AND WHY.  ARE YOUR INTERESTS SIMILAR OR DIFFERENT?   
215596  FIND OUT WHAT THE OTHER CALLER'S FAVORITE TV SHOWS ARE AND WHY.  ARE YOUR INTERESTS SIMILAR OR DIFFERENT?   

        extra  
215593  1      
215594  1      
215595  1      
215596  1

In [167]:
for indx in to_merge:
    old_text = new_df.loc[indx-1, "clean_text"]
    target_text = new_df.loc[indx, "clean_text"].lstrip("--")
    new_df.loc[indx-1, "clean_text"] = old_text.rstrip("--") + target_text  

In [170]:
new_df.iloc[20:30]

dialogue caller  utt_indx  trans_indx  \
20  3774      B      14        21           
21  3774      A      15        22           
22  3774      A      15        23           
23  3774      A      17        25           
24  3774      A      17        26           
25  3774      A      19        28           
26  3774      A      19        29           
27  3774      A      21        31           
28  3774      B      22        32           
29  3774      A      23        33           

                                                                                                                      orig_text  \
20  How about you? /                                                                                                              
21  {D Well } <breathing>, [ we moved in, + when we moved in,  ] [ there, + there ] wasn't any outside lights  /                  
22  {C and --                                                                                                                     
23  -- so } we've been trying to install some, {F uh, } outside lights  /                                                         
24  {C and } --                                                                                                                   
25  -- we put up a fence in the backyard.  /                                                                                      
26  Mostly, {D you know, } not so much thinking that we would deter someone to break in, but that our children would be safe --   
27  -- playing in the yard.                                                                                                       
28  Yeah. /                                                                                                                       
29  {D You know. } /                                                                                                              

   tag  \
20  qo   
21  sd   
22  sd   
23  +    
24  sd   
25  +    
26  sd   
27  +    
28  b    
29  +    

                                                                                                                                   clean_text  \
20  How about you?                                                                                                                              
21  Well, we moved in, when we moved in, there, there wasn't any outside lights                                                                 
22  and  so we've been trying to install some, uh, outside lights                                                                               
23  -- so we've been trying to install some, uh, outside lights                                                                                 
24  and  we put up a fence in the backyard.                                                                                                     
25  -- we put up a fence in the backyard.                                                                                                       
26  Mostly, you know, not so much thinking that we would deter someone to break in, but that our children would be safe  playing in the yard.   
27  -- playing in the yard.                                                                                                                     
28  Yeah.                                                                                                                                       
29  You know.                                                                                                                                   

    topic  \
20  CRIME   
21  CRIME   
22  CRIME   
23  CRIME   
24  CRIME   
25  CRIME   
26  CRIME   
27  CRIME   
28  CRIME   
29  CRIME   

                                                                                                                                             prompt  \
20  DISCUSS CRIME IN AMERICAN CITIES TODAY.  WHAT ARE YOUR CONCE

In [172]:
# теперь можно удалить высказывания с "+"
new_df = new_df.drop(to_merge)
new_df = new_df.reset_index(drop=True)

In [173]:
new_df.head()

dialogue caller  utt_indx  trans_indx  \
0  3774      A      1         0            
1  3774      B      2         1            
2  3774      A      3         2            
3  3774      A      5         5            
4  3774      B      6         6            

                                                                       orig_text  \
0  {C So } I've been concerned about crime lately. /                               
1  Uh-huh. /                                                                       
2  {F Uh, } it's really scary to listen to the news every night  and --            
3  I wondered if you were taking any special precautions in your neighborhood? /   
4  {D Well, } [ I, + I  ] think we have a neighborhood watch <laughter>. /         

    tag  \
0  sd     
1  b      
2  sv     
3  qy^d   
4  na     

                                                                                     clean_text  \
0  So I've been concerned about crime lately.                                                     
1  Uh-huh.                                                                                        
2  Uh, it's really scary to listen to the news every night and  to hear about all the problems.   
3  I wondered if you were taking any special precautions in your neighborhood?                    
4  Well, I, I think we have a neighborhood watch.                                                 

   topic  \
0  CRIME   
1  CRIME   
2  CRIME   
3  CRIME   
4  CRIME   

                                                                                                                                            prompt  \
0  DISCUSS CRIME IN AMERICAN CITIES TODAY.  WHAT ARE YOUR CONCERNS AND THE CONCERNS OF THE OTHER CALLER?  WHAT STEPS CAN BE TAKEN TO REDUCE CRIME?   
1  DISCUSS CRIME IN AMERICAN CITIES TODAY.  WHAT ARE YOUR CONCERNS AND THE CONCERNS OF THE OTHER CALLER?  WHAT STEPS CAN BE TAKEN TO REDUCE CRIME?   
2  DISCUSS CRIME IN AMERICAN CITIES TODAY.  WHAT ARE YOUR CONCERNS AND THE CONCERNS OF THE OTHER CALLER?  WHAT STEPS CAN BE TAKEN TO REDUCE CRIME?   
3  DISCUSS CRIME IN AMERICAN CITIES TODAY.  WHAT ARE YOUR CONCERNS AND THE CONCERNS OF THE OTHER CALLER?  WHAT STEPS CAN BE TAKEN TO REDUCE CRIME?   
4  DISCUSS CRIME IN AMERICAN CITIES TODAY.  WHAT ARE YOUR CONCERNS AND THE CONCERNS OF THE OTHER CALLER?  WHAT STEPS CAN BE TAKEN TO REDUCE CRIME?   

   extra  
0  1      
1  1      
2  1      
3  1      
4  1

## Теперь посмотрим статистику еще раз

In [177]:
len(new_df)

211282

In [178]:
tag_stats = new_df.groupby(by="tag", as_index=False).extra.sum()
tag_stats["extra"] = tag_stats.extra.apply(lambda x: round(x/211282,2))
tag_stats.sort_values(by="extra", ascending=False)[:7]

tag  extra
38  sd  0.36 
11  b   0.16 
39  sv  0.13 
0   %   0.07 
1   +   0.06 
6   aa  0.05 
36  qy  0.02

In [ ]:
# доля b немного уменьшилась, но эти теги нам уже не мешают объединять реплики
# + еще остался. Посмотрим, что это за ситуации

In [179]:
new_df.iloc[:50, :]
# есть случаи, когда несколько плюсов идут подряд, разделенные x
# поэтому сначала удалим x 

dialogue caller  utt_indx  trans_indx  \
0   3774      A      1         0            
1   3774      B      2         1            
2   3774      A      3         2            
3   3774      A      5         5            
4   3774      B      6         6            
5   3774      A      7         7            
6   3774      B      8         8            
7   3774      A      9         9            
8   3774      B      10        10           
9   3774      B      10        11           
10  3774      B      10        12           
11  3774      B      10        13           
12  3774      A      11        14           
13  3774      A      11        15           
14  3774      B      12        16           
15  3774      B      12        17           
16  3774      B      12        18           
17  3774      B      12        19           
18  3774      A      13        20           
19  3774      B      14        21           
20  3774      A      15        22           
21  3774      A      15        23           
22  3774      A      17        26           
23  3774      A      19        29           
24  3774      B      22        32           
25  3774      A      23        33           
26  3774      B      24        34           
27  3774      A      25        35           
28  3774      B      26        37           
29  3774      A      27        38           
30  3774      A      31        43           
31  3774      B      34        46           
32  3774      A      35        47           
33  3774      B      36        48           
34  3774      A      37        49           
35  3774      A      39        52           
36  3774      B      40        54           
37  3774      B      40        55           
38  3774      A      41        56           
39  3774      B      42        57           
40  3774      B      42        58           
41  3774      A      43        59           
42  3774      B      44        60           
43  3774      B      44        61           
44  3774      B      46        63           
45  3774      B      48        65           
46  3774      B      48        66           
47  3774      A      49        67           
48  3774      A      49        68           
49  3774      A      49        69           

                                                                                                                           orig_text  \
0   {C So } I've been concerned about crime lately. /                                                                                  
1   Uh-huh. /                                                                                                                          
2   {F Uh, } it's really scary to listen to the news every night  and --                                                               
3   I wondered if you were taking any special precautions in your neighborhood? /                                                      
4   {D Well, } [ I, + I  ] think we have a neighborhood watch <laughter>. /                                                            
5   Uh-huh. /                                                                                                                          
6   I think. /                                                                                                                         
7   <Laughter>.                                                                                                                        
8   I'm not real, - /                                                                                                                  
9   we don't get real involved.  /                                                                                                     
10  We're never home,  /                                                                                                               
11  {C so } -- -/                                                           

In [181]:
to_delete = []
counter = 0

for i in new_df.index:
    if new_df.loc[i, "tag"] == "x":
        counter += 1
        to_delete.append(i)

In [182]:
counter

3630

In [184]:
to_delete[:5]

[7, 13, 31, 33, 58]

In [186]:
new_df[new_df.tag == "x"][:5]

dialogue caller  utt_indx  trans_indx           orig_text tag clean_text  \
7   3774      A      9         9           <Laughter>.         x               
13  3774      A      11        15          <laughter>.         x   .           
31  3774      B      34        46          <Laughter>.         x               
33  3774      B      36        48          <Laughter>.         x               
58  3774      A      55        78          <Throat_clearing>.  x               

    topic  \
7   CRIME   
13  CRIME   
31  CRIME   
33  CRIME   
58  CRIME   

                                                                                                                                             prompt  \
7   DISCUSS CRIME IN AMERICAN CITIES TODAY.  WHAT ARE YOUR CONCERNS AND THE CONCERNS OF THE OTHER CALLER?  WHAT STEPS CAN BE TAKEN TO REDUCE CRIME?   
13  DISCUSS CRIME IN AMERICAN CITIES TODAY.  WHAT ARE YOUR CONCERNS AND THE CONCERNS OF THE OTHER CALLER?  WHAT STEPS CAN BE TAKEN TO REDUCE CRIME?   
31  DISCUSS CRIME IN AMERICAN CITIES TODAY.  WHAT ARE YOUR CONCERNS AND THE CONCERNS OF THE OTHER CALLER?  WHAT STEPS CAN BE TAKEN TO REDUCE CRIME?   
33  DISCUSS CRIME IN AMERICAN CITIES TODAY.  WHAT ARE YOUR CONCERNS AND THE CONCERNS OF THE OTHER CALLER?  WHAT STEPS CAN BE TAKEN TO REDUCE CRIME?   
58  DISCUSS CRIME IN AMERICAN CITIES TODAY.  WHAT ARE YOUR CONCERNS AND THE CONCERNS OF THE OTHER CALLER?  WHAT STEPS CAN BE TAKEN TO REDUCE CRIME?   

    extra  
7   1      
13  1      
31  1      
33  1      
58  1

In [187]:
new_df = new_df.drop(to_delete)
new_df = new_df.reset_index(drop=True)

## Теперь надо придумать, что делать с другими "плюсами"
## Во избежание ошибок пока просто проставим им предыдущий тег того же спикера

In [196]:
new_df_copy = new_df.copy()

In [198]:
def replace_plus(df: pd.DataFrame, i:int, j:int):
    speaker = df.loc[i, "caller"]
    if df.loc[j-1, "caller"] == speaker:
        df.loc[i, "tag"] = df.loc[j-1, "tag"]
    else:
        j -= 1
        return replace_plus(df, i, j)
    

for i in tqdm(new_df_copy.index):
    if new_df_copy.loc[i, "tag"] == "+":      
        j = i
        replace_plus(new_df_copy, i, j)

In [207]:
# вроде бы все получилось

# new_df.iloc[0:50, :]

In [206]:
# new_df_copy.iloc[0:50, :]

In [235]:
new_df_copy.head(14)

dialogue caller  utt_indx  trans_indx  \
0   3774      A      1         0            
1   3774      B      2         1            
2   3774      A      3         2            
3   3774      A      5         5            
4   3774      B      6         6            
5   3774      A      7         7            
6   3774      B      8         8            
7   3774      B      10        10           
8   3774      B      10        11           
9   3774      B      10        12           
10  3774      B      10        13           
11  3774      A      11        14           
12  3774      B      12        16           
13  3774      B      12        17           

                                                                             orig_text  \
0   {C So } I've been concerned about crime lately. /                                    
1   Uh-huh. /                                                                            
2   {F Uh, } it's really scary to listen to the news every night  and --                 
3   I wondered if you were taking any special precautions in your neighborhood? /        
4   {D Well, } [ I, + I  ] think we have a neighborhood watch <laughter>. /              
5   Uh-huh. /                                                                            
6   I think. /                                                                           
7   I'm not real, - /                                                                    
8   we don't get real involved.  /                                                       
9   We're never home,  /                                                                 
10  {C so } -- -/                                                                        
11  Uh-huh  /                                                                            
12  -- {F uh, } <lipsmack> {F uh, } {D well, } I know they were going to start one,  /   
13  {C but, } {F uh, } I haven't heard any more since,  /                                

     tag  \
0   sd     
1   b      
2   sv     
3   qy^d   
4   na     
5   b      
6   sd     
7   %      
8   sd     
9   sd     
10  %      
11  b      
12  sd     
13  sd     

                                                                                      clean_text  \
0   So I've been concerned about crime lately.                                                     
1   Uh-huh.                                                                                        
2   Uh, it's really scary to listen to the news every night and  to hear about all the problems.   
3   I wondered if you were taking any special precautions in your neighborhood?                    
4   Well, I, I think we have a neighborhood watch.                                                 
5   Uh-huh.                                                                                        
6   I think.                                                                                       
7   I'm not real,                                                                                  
8   we don't get real involved.                                                                    
9   We're never home,                                                                              
10  so --                                                                                          
11  Uh-huh                                                                                         
12  -- uh, uh, well, I know they were going to start one,                                          
13  but, uh, I haven't heard any more since,                                                       

    topic  \
0   CRIME   
1   CRIME   
2   CRIME   
3   CRIME   
4   CRIME   
5   CRIME   
6   CRIME   
7   CRIME   
8   CRIME   
9   CRIME   
10  CRIME   
11  CRIME   
12  CRIME   
13  CRIME   

                                                                                                   

In [ ]:
# потом надо написать правило, что если между одинаковыми единицами стоит b, то игнорируем b и мерджим все одинаковые теги

## Теперь надо решить, что делать с "%" - этот тег означает, что сложно понять, что это за речевой акт 
Наверное, можно оставить его как есть, но потом не учитывать

## Попробуем объединить данные из датасета в целые реплики и теги по спикерам

In [225]:
# sample_df = new_df_copy.iloc[0:200, :]

In [275]:
grouped = new_df_copy.groupby(by="dialogue")

In [276]:
d = {}

for name,group in tqdm(grouped):
    dialogue_data = []
    speaker_data = []
    for i in group.index:
        if speaker_data == []:
            text = group.loc[i, "clean_text"].strip("--")
            tag = group.loc[i, "tag"]
            speaker_data.append((text, tag))
        else:
            if group.loc[i,"caller"] == group.loc[i-1,"caller"]:
                text = group.loc[i, "clean_text"].strip("--")
                tag = group.loc[i, "tag"]
                speaker_data.append((text, tag))
            else:
                dialogue_data.append(speaker_data)
                speaker_data = []
                text = group.loc[i, "clean_text"].strip("--")
                tag = group.loc[i, "tag"]
                speaker_data.append((text, tag))
    d[name] = dialogue_data

In [277]:
len(d)

1155

In [278]:
d[3774][:7]

[[("So I've been concerned about crime lately.", 'sd')],
 [('Uh-huh.', 'b')],
 [("Uh, it's really scary to listen to the news every night and  to hear about all the problems.",
   'sv'),
  ('I wondered if you were taking any special precautions in your neighborhood?',
   'qy^d')],
 [('Well, I, I think we have a neighborhood watch.', 'na')],
 [('Uh-huh.', 'b')],
 [('I think.', 'sd'),
  ("I'm not real,", '%'),
  ("we don't get real involved.", 'sd'),
  ("We're never home,", 'sd'),
  ('so ', '%')],
 [('Uh-huh', 'b')]]

Вроде бы все ок

## Теперь объединим одинаковые теги в подряд идущих тегах одного спикера и, соответственно, их тексты (игнорируя тег %).

In [279]:
for k, talk in d.items():
    new_turns = []
    for turn in talk:
        # unzip lists of tuples
        turn = [[i for i, j in turn], [j for i, j in turn]] 
        new_turns.append(turn)
    d[k] = new_turns

In [280]:
d[3774][:7]

[[["So I've been concerned about crime lately."], ['sd']],
 [['Uh-huh.'], ['b']],
 [["Uh, it's really scary to listen to the news every night and  to hear about all the problems.",
   'I wondered if you were taking any special precautions in your neighborhood?'],
  ['sv', 'qy^d']],
 [['Well, I, I think we have a neighborhood watch.'], ['na']],
 [['Uh-huh.'], ['b']],
 [['I think.',
   "I'm not real,",
   "we don't get real involved.",
   "We're never home,",
   'so '],
  ['sd', '%', 'sd', 'sd', '%']],
 [['Uh-huh'], ['b']]]

In [304]:
# теперь объединяем тексты с одинаковыми тегами

new_d = {}

for k, talk in d.items():
        
    new_turns = []
    
    for turn in talk:
        
        new_texts = []
        new_tags = []
        
        if len(turn[0]) == 1:
            new_turns.append(turn)
        
        if len(turn[0]) > 1:
            
            cur_texts = turn[0]
            cur_tags = turn[1]
        
            for i, tag in enumerate(cur_tags):
                if i == 0:
                    new_texts.append(cur_texts[i])
                    new_tags.append(cur_tags[i])
                else:
                    if cur_tags[i] == new_tags[-1]:
                        new_texts[-1] += " " + cur_texts[i]
                    else:
                        new_texts.append(cur_texts[i])
                        new_tags.append(cur_tags[i])
                    
        if new_texts != []:
            new_turns.append([new_texts, new_tags])
        
    new_d[k] = new_turns 

In [305]:
d[3774][:7]

[[["So I've been concerned about crime lately."], ['sd']],
 [['Uh-huh.'], ['b']],
 [["Uh, it's really scary to listen to the news every night and  to hear about all the problems.",
   'I wondered if you were taking any special precautions in your neighborhood?'],
  ['sv', 'qy^d']],
 [['Well, I, I think we have a neighborhood watch.'], ['na']],
 [['Uh-huh.'], ['b']],
 [['I think.',
   "I'm not real,",
   "we don't get real involved.",
   "We're never home,",
   'so '],
  ['sd', '%', 'sd', 'sd', '%']],
 [['Uh-huh'], ['b']]]

In [306]:
# вот что получается
new_d[3774][:7]

[[["So I've been concerned about crime lately."], ['sd']],
 [['Uh-huh.'], ['b']],
 [["Uh, it's really scary to listen to the news every night and  to hear about all the problems.",
   'I wondered if you were taking any special precautions in your neighborhood?'],
  ['sv', 'qy^d']],
 [['Well, I, I think we have a neighborhood watch.'], ['na']],
 [['Uh-huh.'], ['b']],
 [['I think.',
   "I'm not real,",
   "we don't get real involved. We're never home,",
   'so '],
  ['sd', '%', 'sd', '%']],
 [['Uh-huh'], ['b']]]

In [ ]:
# осталось разобраться с % 

## Пока сохраню new_df_copy и текущую версию словаря диалогов

In [308]:
joblib.dump(new_df_copy, "corpus-df.pkl")
joblib.dump(new_d, "corpus-dict.pkl")

['corpus-dict.pkl']

In [222]:
print(sorted(df.tag.unique()))

['%', '+', '^2', '^g', '^h', '^q', 'aa', 'aap_am', 'ad', 'ar', 'arp_nd', 'b', 'b^m', 'ba', 'bd', 'bf', 'bh', 'bk', 'br', 'fa', 'fc', 'fo_o_fw_"_by_bc', 'fp', 'ft', 'h', 'na', 'ng', 'nn', 'no', 'ny', 'oo_co_cc', 'qh', 'qo', 'qrr', 'qw', 'qw^d', 'qy', 'qy^d', 'sd', 'sv', 't1', 't3', 'x']


Все теги:
  q      Question  
  s      Statement 
  b      Backchannel/Backwards-Looking
  f      Forward-Looking
  a      Agreements
  %  indeterminate, interrupted, or contains just a floor holder (see manual)
  (^u  [on anything] unrelated response (first utt is NOT response to previous q)
  *  comment  (followed by "*[[comment...]]" after transcription to explain)
  +  continued from previous by same speaker
  @,o@,+@  incorrect transcription (can add comment to specify problem further)
  ^2 collaborative completion
  ^c  about-communication
  ^d  declarative question (question asked like a structural statement)
  ^e  [on statements] elaborated reply to y/n question
  ^g  tag question (question asked like a structural statement with a question tag at end)
  ^h  hold (often but not always after a question) ('let me think'; question in response to a question)
  ^m  mimic other
  ^q  quotation
  ^r  repeat self
  ^t  about-task
  aap Accept-part    
  ad Action-directive  "Go ahead", "We could go back to television shows"
  aa Accept         "ok" , "i agree"
  am Maybe                         
  ar Reject "no", 
  arp Reject-part 
  b default agreement or continuer (uh-huh, right, yeah)
  b^m  Repeat-phrase  
  ba assessment/appreciation ("I can imagine")
  bc Correct-misspeaking  
  bd Downplaying-reponse-to-sympathy/compliments ("That's all right","that happens")
  bf reFormulate/summarize; paraphrase/summary of other's utterance (as opposed to a mimic)
  bh rhetorical question continuer ("Oh really?")
  bk ACKNOWLEDGE-ANSWER    "Oh, okay"
  br Signal-non-understanding (request for repeat)
  br^m Signal-non-understanding via mimic
  br^c non-understanding due to problems with phone line  
  by sYmpathetic comment ("I'm sorry to hear about that")
  cc Commit                          
  co Offer                           
  fa Apology "Apologies" (this is not the "I'm sorry" of sympathy which is "by")
  fc Conventional-closing            
  fe Exclamation "Ouch"
  fo Other-forward-function         
  fp Conventional-opening            
  ft Thanks "Thank you"
  fw Welcome "You're welcome"
  fx Explicit-performative  ("you're filed" )      
  na a descriptive/narrative statement which acts as an affirmative answer to a question 
  nd aNswer Dispreferred (Well...)
  ng a descriptive/narrative statement which acts as a negative answer to a question 
  nn  no or variations (only)
  no a response to a question that is neither affirmative nor negative (often "I don't know")
  ny  yes or variations (only)
  o other
  oo Open-option  "We could have lamb or chicken"
  qh  rhetorical question
  qo  open ended question
  qr  alternative ('or') question 
  qrr an or-question clause tacked onto a yes-no question
  qw  wh-question 
  qy  yes/no question
  sd  descriptive and/or narrative (listener has no basis to dispute)
  sv  viewpoint, from personal opinions to proposed general facts  (listener could have basis to dispute)
  t1  self-talk
  t3  3rd-party-talk
  x   nonspeech 